### PR TITLE
pkg/lwip: add support for HAL radios that require IRQ offloading

### PR DESCRIPTION
**Pull request by jia200x**
### Contribution description
This PR utilizes https://github.com/RIOT-OS/RIOT/pull/18464 to run radio HAL based devices with LWIP. This also allows the usage of other non-netdev devices that require ISR offloading.

As a result of this PR, I also push support for KW2XRF radios on LWIP.

### Testing procedure
Murdock compile output should be fine
Test LWIP with any KW2XRF based radio.

### Issues/PRs references
~Depends on https://github.com/RIOT-OS/RIOT/pull/18464~
Now it depends on https://github.com/RIOT-OS/RIOT/pull/18479

### CODE REVIEW

**Comment by benpicco** [comment]
Why can't LWIP use `bhp_event`?

**Comment by jia200x** [comment]
It definitely can, I just wanted to speed up the migration process of the HAL. I was planning anyway to migrate all pending IPC based IRQ offloaders at some point.

**Comment by benpicco** [comment]
This needs a rebase

**Comment by jia200x** [comment]
rebased!

File: pkg/lwip/init_devs/auto_init_kw2xrf.c [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+static lwip_netif_t netif[NETIF_KW2XRF_NUMOF];
+static kw2xrf_t kw2xrf_devs[NETIF_KW2XRF_NUMOF];
+static netdev_ieee802154_submac_t kw2xrf_netdev[NETIF_KW2XRF_NUMOF];
+
+static void auto_init_kw2xrf(void)
+{
+    for (unsigned i = 0; i < NETIF_KW2XRF_NUMOF; i++) {
+        bhp_msg_init(&netif[i].bhp, &kw2xrf_radio_hal_irq_handler, &kw2xrf_netdev[i].submac.dev);
+        kw2xrf_init(&kw2xrf_devs[i], &kw2xrf_params[i], &kw2xrf_netdev[i].submac.dev,
+                        bhp_msg_isr_cb, &netif[i].bhp);
```
Comment (benpicco) [inline pkg/lwip/init_devs/auto_init_kw2xrf.c:42]:
```suggestion
        kw2xrf_init(&kw2xrf_devs[i], &kw2xrf_params[i], &kw2xrf_netdev[i].submac.dev,
                    bhp_msg_isr_cb, &netif[i].bhp);
```

**Review by benpicco — COMMENTED** [review]
Looks good to me - there is now a leftover `kw2xrf_setup()` in `kw2xrf.h` - want to get rid of that while you're at it?

File: pkg/lwip/init_devs/auto_init_kw2xrf.c [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+        netdev_register(&kw2xrf_netdev[i].dev.netdev, NETDEV_KW2XRF, i);
```
Comment (benpicco) [inline pkg/lwip/init_devs/auto_init_kw2xrf.c:44]: Why is this not done by `kw2xrf_init()`?

Comment (jia200x) [reply pkg/lwip/init_devs/auto_init_kw2xrf.c:44]: because that would add a dependency to `netdev` that does not exist in the current Radio HAL drivers.

**Comment by jia200x** [comment]
> Looks good to me - there is now a leftover `kw2xrf_setup()` in `kw2xrf.h` - want to get rid of that while you're at it?

sure, good catch. I can remove it

**Comment by jia200x** [comment]
done!

**Review by benpicco — APPROVED** [review]
Code looks good, I'm trusting your testing

**Comment by jia200x** [comment]
hmmmm I found an assertion on the KW2XRF on LWIP. I'm investigating the problem.

**Comment by jia200x** [comment]
ouch... although the majority of drivers were designed to run on a single thread, LWIP uses a dedicated thread for performing IRQ offloading :/. This breaks the assumption of the SubMAC that everything runs on a single thread.
This might not be that easy to fix, since `lwip_sock` sends from the caller thread.

I will see what I can do.

**Comment by jia200x** [comment]
Probably a mutex can do the trick, but still, this should probably be handled in a separate PR, as it affects essentially all network devices

**Comment by jia200x** [comment]
It uses only only uses "one thread" (with quotes because it actually allocates one for every network device), everything else runs in the context of the caller (for `tests/lwip` this would be `main`).
Due to the nature of the sock API we need "at least" an extra thread, similar to GNRC. The problem lies on concurrency between `netdev` calls from different threads.

I just implemented a version with `rmutex` (although it's not my favourite sync mechanism) and it seems to work fine.

IMO the solutions are either:
1. We stick to these recursive mutexes.
2. We ensure all netdev calls occur on the same IRQ offloader thread. Since the IRQ offload thread always runs with the highest priority, we could just send a message to the thread with the right context.

Any preference?

**Comment by maribu** [comment]
I think it is possible to use a single thread to handle both lwip events as well as ISRs. With `sys/event`, we can event wait on multiple queues to implement priorities. Likely it does make sense to boost the priority of ISR events, as some drives have to deadlines to fullfil (e.g. timeouts for ACKs, or the CC110x that is writing frames of up to 255 bytes into a 64 byte fifo under the assumption the MCU can drain the FIFO faster than the transceiver can fill it).

But having an rmutex until then sounds good to me.

**Comment by benpicco** [comment]
This needs a rebase, but then it should be good to merge, right?

**Comment by jia200x** [comment]
rebased! and yes, it should be ready to merge

**Comment by benpicco** [comment]
CI is unhappy

**Comment by jia200x** [comment]
> CI is unhappy

Unrelated error

**Comment by MrKevinWeiss** [comment]
I really hope I fixed all the unrelated error issues...

**Comment by MrKevinWeiss** [comment]
try building locally [`examples/paho-mqtt` with `arduino-mkrwan1300` or other boards](https://ci.riot-os.org/RIOT-OS/RIOT/18465/c3e4fa13a7df36a24116e95e72990a3de9d9fbf5/output/compile/examples/paho-mqtt/arduino-mkrwan1300:gnu.txt)

**Comment by benpicco** [comment]
If you look at the [error](https://ci.riot-os.org/RIOT-OS/RIOT/18465/c3e4fa13a7df36a24116e95e72990a3de9d9fbf5/output/compile/examples/paho-mqtt/dwm1001:gnu.txt) [output](https://ci.riot-os.org/RIOT-OS/RIOT/18465/c3e4fa13a7df36a24116e95e72990a3de9d9fbf5/output/compile/examples/paho-mqtt/arduino-mkrwan1300:gnu.txt) it is clearly related to this PR.

**Comment by jia200x** [comment]
it was a rebase error. Now it should be fixed

**Pull request merged**